core/modules/wcontrol/TransmitMgr.h

```diff
@@ -62,51 +62,72 @@ class TransmitMgr {
     TransmitMgr& operator=(TransmitMgr const&) = delete;
     virtual ~TransmitMgr() = default;
 
-    int getTotalCount() { return _totalCount; }
-    int getTransmitCount() { return _transmitCount; }
-    int getAlreadyTransCount() { return _alreadyTransCount; }
+    int getTotalCount(int czarId);
+    int getTransmitCount(int czarId);
+    int getAlreadyTransCount(int czarId);
 
-    virtual std::ostream& dump(std::ostream &os) const;
-    std::string dump() const;
+    /// Class methods, that have already locked '_mtx', should call 'dumpBase'.
+    std::ostream& dump(std::ostream &os) const;
+
+    /// This will try to lock 'TransmitMgr::_mtx'.
     friend std::ostream& operator<<(std::ostream &out, TransmitMgr const& mgr);
 
     friend class TransmitLock;
 
+protected:
+    /// _mtx must be locked before calling this function.
+    /// Dump the contents of the class to a string for logging.
+    virtual std::ostream& dumpBase(std::ostream &os) const;
+
+    /// _mtx must be locked before calling this function.
+    std::string dump() const;
+
 private:
-    void _take(bool interactive, bool alreadyTransmitting);
+    void _take(bool interactive, bool alreadyTransmitting, int czarId);
+
+    void _release(bool interactive, bool alreadyTransmitting, int czarId);
+
+    /// This class is used to store transmit information for a czar
+    class TransmitInfo {
+        using Ptr = std::shared_ptr<TransmitInfo>;
```

Comment:
The class has tiny footprint (3 primitive data members). I believe it makes very little sense to handle objects of the class via `shared_ptr` rather than passing them by values. The pointer management would just complicate things w/o any added benefits.

Reply:
It is possible for the TransmitMgr::_release function to see a value of zero for at least _transmitCount if TransmitMgr::_take is waiting. I'm using the shared pointer to tell if _take is waiting before erasing the item. I can probably use a flag in TransmitInfo for the same purpose.